packages/eslint-plugin/src/rules/switch-exhaustiveness-check.ts

```diff
@@ -0,0 +1,129 @@
+import { TSESLint, TSESTree } from '@typescript-eslint/experimental-utils';
+import * as ts from 'typescript';
+import {
+  createRule,
+  getParserServices,
+  getConstrainedTypeAtLocation,
+} from '../util';
+import { isTypeFlagSet, unionTypeParts } from 'tsutils';
+
+export default createRule({
+  name: 'switch-exhaustiveness-check',
+  meta: {
+    type: 'suggestion',
+    fixable: 'code',
+    docs: {
+      description: 'Exhaustiveness checking in switch with union type',
+      category: 'Best Practices',
+      recommended: false,
+      requiresTypeChecking: true,
+    },
+    schema: [],
+    messages: {
+      switchIsNotExhaustive:
+        'Switch is not exhaustive. Cases not matched: {{missingBranches}}',
+      addMissingCases: 'Add branches for missing cases',
+    },
+  },
+  defaultOptions: [],
+  create(context) {
+    const service = getParserServices(context);
+    const checker = service.program.getTypeChecker();
+
+    function getNodeType(node: TSESTree.Node): ts.Type {
+      const tsNode = service.esTreeNodeToTSNodeMap.get(node);
+      return getConstrainedTypeAtLocation(checker, tsNode);
+    }
+
+    function fixSwitch(
+      fixer: TSESLint.RuleFixer,
+      node: TSESTree.SwitchStatement,
+      missingBranchTypes: Array<ts.Type>,
+    ): TSESLint.RuleFix | null {
+      // It seems, there is no way to insert automatically generated cases
+      // inside an empty switch body
+      if (node.cases.length === 0) {
+        return null;
+      }
```

Comment:
you should be able to do this, you'll just have to branch your logic slightly.

Example:
https://astexplorer.net/#/gist/24cef4bf01c70867f050b0365a54f894/2f19c607f988e878a733d159839eb5ad0393527c
```ts
import { isClosingBraceToken, isOpeningBraceToken } from 'eslint-utils';

function fixSwitch(fixer, node, missingBranchTypes) {
  const lastCase = node.cases.length > 0 ? node.cases[node.cases.length - 1] : null;
  const lastCaseIndent = lastCase
    ? Array(lastCase.loc.start.column + 1).join(' ')
    // if there's no cases, just infer no indent and make the user format it correctly
    : '';

  // ... existingCode ...

  if (lastCase) {
    return fixer.insertTextAfter(lastCase, fixString);
  }
 
  // there were no existing cases
  const openingBrace = sourceCode.getTokenAfter(node.discriminant, isOpeningBraceToken);
  const closingBrace = sourceCode.getTokenAfter(node.discriminant, isClosingBraceToken);

  return fixer.replaceTextRange([openingBrace.range[0], closingBrace.range[1]], [
    '{',
    fixString,
    '}',
  ].join('\n'))
}
```

Reply:
Thanks for the pointers on this!

I've made this change, but after looking at suggestions for when there are no cases I decided to use `switch` indent instead of none, e.g.:

```
    switch (value) {
case 1: { throw new Error('Not implemented yet: 1 case') }
case 2: { throw new Error('Not implemented yet: 2 case') }
}
```

vs

```
    switch (value) {
    case 1: { throw new Error('Not implemented yet: 1 case') }
    case 2: { throw new Error('Not implemented yet: 2 case') }
    }
```

I know we want to avoid formatting anything here and pass that responsibility to the user, but I found that that closing brace without indentation looks way too misleading, as if the top level block is being closed. Let me know if you don't think it's a concern.